### ISSUE TITLE
[MNG-6648] 'mavenrc_pre' script does not receive arguments like mavenrc in Bourne shell does

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8577]
**[Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd)** opened **[MNG-6648](https://issues.apache.org/jira/browse/MNG-6648?redirect=false)** and commented

Hello,

In some corner-case, you could have to run Maven with a custom `JAVA_HOME` (JDK v8, v11, ... even if it is evil some plugin can be impacted by JDK at runtime).

If it is per-project, [JVM and Command Line Options](https://maven.apache.org/docs/3.3.1/release-notes.html#JVM_and_Command_Line_Options) is not relevant and update `JAVA_HOME` each time in your environment could be boring.

A command line user-custom-parameter (ex : `-Dmy.jdk=11`) could be simple to achieve that.

A Maven script wrapper could be implemented for that, but it's like update the Maven distribution => evil.
The **mavenrc_pre** (MNGSITE-246) could be perfect for that feature !
But currently command line parameters are not used in script :(.

If parameters are used dor script call, sample for `mvn.cmd` (windows):

```
if exist "%USERPROFILE%\mavenrc_pre.cmd" call "%USERPROFILE%\mavenrc_pre.cmd %*"
```

It's allow any user customization, using Maven standards.

If relevant, I can provide PR.
Best regards

---

**Affects:** 3.6.0, 3.6.1

**Issue Links:**
- [MWRAPPER-27](https://issues.apache.org/jira/browse/MWRAPPER-27) rework mvnw* scripts to better match mvn* from Maven 3.8

**Remote Links:**
- [GitHub Pull Request #511
](https://github.com/apache/maven/pull/511)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

1 votes, 4 watchers

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

You can simply define JAVA_HOME in `mavenrc_pre.cmd` ?

```
SET JAVA_HOME=C:\....
```

Furthermore you can use toolchains? Can you elaborate more in detail what the real problem is?

**Comment by jira-importer** [comment]
**[Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd)** commented

Hello [Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise),

As my understanding, the toolchains can customize the JDK/JRE used by plugins (javac for compiler, ...), but not the JRE used at runtime for Maven, and oblige to update the pom.

My use case is to customize the Maven runtime JRE by command line parameter (some rare plugins could again have a behavior linked to Maven runtime JRE), allowing to switch this JRE simply by project (it is for corner-case use-case, but it's exist when you develop on very-old and new projects).

In addition, it could help to upgrade Maven version when you have some historical wrapper to achieve "by project behavior" by parameters on command line (coming from Maven 2 usage and custom Java version & memory settings needs by project behavior ; when per-project features doesn't exist, implemented in 3.3).

So interpret command lines parameters to check/verify/enforce custom items and back to Maven standards configuration could be helpful, feasible by design with `mavenrc_pre` if commande line parameters are passed to this script.

Best regards

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi, could you make a concrete example of your problem? Furthermore `mavenre_pre` to get command line parameters to do some checking? It's not intended there..? The question is what kind of checks/verify/enforce you have in mind?
Furthermore why not giving `JAVA_HOME` via command line? Like `set JAVA_HOME=...;mvn clean` would work in Linux/MacOs etc ? I don't know if something like this exists for Windows?

**Comment by jira-importer** [comment]
**[Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd)** commented

The use-case is (for sample) to use JDK 11 if enforced by command line parameter, or JDK 1.8 if JAVA_HOME empty.

Command line: `mvn -Dmaven.jdk=11 clean`

Could be done with a script like:

```
@echo off

rem Define JDK 1.7 if enforced as command line parameter
echo "%*" | findstr "maven.jdk=7" >NUL
IF %ERRORLEVEL% EQU 0 (set JAVA_HOME=c:\somewhere\jdk1.7)

rem Define JDK 11 if enforced as command line parameter
echo "%*" | findstr "maven.jdk=11" >NUL
IF %ERRORLEVEL% EQU 0 (set JAVA_HOME=c:\somewhere\jdk11)

rem If JAVA_HOME empty, force to JDK 1.8 (avoiding using "C:\Program Files (x86)\Java" if dos-cmd or the one of IDE runtime)
IF "%JAVA_HOME%" == "" (set JAVA_HOME=c:\somewhere\jdk1.8)

rem Debug for result check
echo Current JAVA_HOME: %JAVA_HOME%
```

I agree that `set JAVA_HOME=c:\somewhere\jdkXX && mvn clean` is working on Windows, but when Maven used in command line.
If you bind a Maven external distribution in your IDE and execute it with integrated runner, you should have to custom dedicated runner (sysenv properties, ...) to enforce JDK.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This case is perfectly valid, actually this is a bug. In the Bourne shell script we do source `mavenrc` which is part of the `mvn` script now and has full access to `"$@"`. In Windows Command Prompt there is no equivalent to the source commmand. Therefore, we use `call`, but shall is rather similiar to invoking another script and that script won't have access to the arguments passed to `mvn.cmd`. For that reason [Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd), is right, we need to do:

```
if exist "%USERPROFILE%\mavenrc_pre.bat" call "%USERPROFILE%\mavenrc_pre.bat" %*
if exist "%USERPROFILE%\mavenrc_pre.cmd" call "%USERPROFILE%\mavenrc_pre.cmd" %*
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [f7a775b79a195b5649c8c0309b673a4d5bdcf1c1](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=f7a775b79a195b5649c8c0309b673a4d5bdcf1c1) and [22a8cfa05976b23284388675dcabd7e5e8037fb0](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=22a8cfa05976b23284388675dcabd7e5e8037fb0) for `maven-3.8.x` branch.
